manager/state/raft/transport/transport.go

```diff
@@ -0,0 +1,250 @@
+
+type TransportRaft interface {
+	Process(ctx context.Context, m raftpb.Message)
+	ReportUnreachable(id uint64)
+	ReportSnapshot(id uint64, status raft.SnapshotStatus)
+	IsIDRemoved(id uint64) bool
+
+	NodeRemoved()
+}
+
+type Config struct {
+	SendTimeout time.Duration
+	Credentials credentials.TransportCredentials
+
+	TransportRaft TransportRaft
+}
+
+type Transport struct {
+	config *Config
+
+	mu    sync.RWMutex
+	peers map[uint64]*peer
+
+	stopped bool
+
+	done chan struct{}
+}
+
+func New(cfg *Config) *Transport {
+	return &Transport{
+		peers:  make(map[uint64]*peer),
+		config: cfg,
+		done:   make(chan struct{}),
+	}
+}
+
+func (t *Transport) Run(ctx context.Context) {
+	<-ctx.Done()
+	log.G(ctx).Debug("stop transport")
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.stopped = true
+	for _, p := range t.peers {
+		p.stop()
+	}
+	close(t.done)
+}
+
+func (t *Transport) Done() chan struct{} {
+	return t.done
+}
+
+func (t *Transport) AddPeer(ctx context.Context, id uint64, addr string) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.stopped {
+		return errors.New("transport stopped")
+	}
+	if ep, ok := t.peers[id]; ok {
+		return errors.Errorf("peer %x already added with addr %s", id, ep.addr)
+	}
+	p, err := newPeer(id, addr, t)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create peer %x with addr %s", id, addr)
+	}
+	t.peers[id] = p
+	go p.processLoop(ctx)
+	return nil
+}
+
+func (t *Transport) RemovePeer(id uint64) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	p, ok := t.peers[id]
+	if !ok {
+		return ErrIsNotFound
+	}
+	p.stop()
+	delete(t.peers, id)
+	return nil
+}
+
+func (t *Transport) UpdatePeer(ctx context.Context, id uint64, addr string) error {
+	t.mu.Lock()
+	p, ok := t.peers[id]
+	t.mu.Unlock()
+	if !ok {
+		return ErrIsNotFound
+	}
+	if addr == p.address() {
+		return nil
+	}
+	cc, err := t.dial(addr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create conn for %x with addr %s", id, addr)
+	}
+	ctx, cancel := context.WithTimeout(ctx, t.config.SendTimeout)
+	defer cancel()
+	resp, err := api.NewHealthClient(cc).Check(ctx, &api.HealthCheckRequest{Service: "Raft"})
+	if err != nil {
+		cc.Close()
+		return errors.Wrap(err, "failed to check health")
+	}
+	if resp.Status != api.HealthCheckResponse_SERVING {
+		cc.Close()
+		return fmt.Errorf("health check returned status %s", resp.Status)
+	}
+	p.update(cc, addr)
+	return nil
+}
+
+func (t *Transport) GetPeerAddr(id uint64) (string, error) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	p, ok := t.peers[id]
+	if !ok {
+		return "", ErrIsNotFound
+	}
+	return p.address(), nil
+}
+
+func (t *Transport) sendTemporary(ctx context.Context, m raftpb.Message) error {
+	longestActive, err := t.longestActive()
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(ctx, 2*t.config.SendTimeout)
+	defer cancel()
+	addr, err := longestActive.resolveAddr(ctx, m.To)
+	if err != nil {
+		return err
+	}
+	p, err := newPeer(m.To, addr, t)
```

Comment:
It's problematic to use a peer that isn't tracked in the `t.peers` list. For example, that peer might call `ReportUnreachable` after raft has shut down.